batchflow/models/torch/base_batchsize_mixin.py

```diff
@@ -0,0 +1,336 @@
+
+            self.garbage_collection_cuda()
+            if exit or count >= max_iters:
+                break
+
+        return batch_size[-1]
+
+
+    def _compute_optimal_batch_size_predictive(self, method='train', max_memory=90,
+                                               inputs=None, targets=None, pbar='n',
+                                               max_iters=16, start_batch_size=4,
+                                               delta_batch_size=4, max_batch_size=128,
+                                               n=20, frequency=0.05, time_threshold=3,
+                                               tail_size=20, std_threshold=0.1):
+        """ Compute optimal batch size for training/inference to maximize GPU memory usage.
+        Works by using `train`/`predict` with different batch sizes, and measuring how much memory is taken.
+        Then, we solve the system of `measured_memory = batch_size * item_size + model_size + eps` equations for both
+        `item_size` and `model_size`.
+
+        For stable measurements, we make `n` iterations of `train`/`predict`, until the memory consumption stabilizes.
+        """
+        #pylint: disable=consider-iterating-dictionary
+        table = {}
+        batch_size = start_batch_size
+        for _ in Notifier(pbar)(range(max_iters)):
+            info = self.get_memory_utilization(batch_size, method=method,
+                                               inputs=inputs, targets=targets, n=n, frequency=frequency,
+                                               time_threshold=time_threshold,
+                                               tail_size=tail_size, std_threshold=std_threshold)
+            table[batch_size] = info
+
+            # Exit condition
+            batch_size += delta_batch_size
+            if info['memory'] > max_memory or batch_size > max_batch_size:
+                break
+
+        # Make and solve a system of equations for `item_size`, `model_size`
+        matrix = np.array([[batch_size, 1] for batch_size in table.keys()])
+        vector = np.array([value['memory'] for value in table.values()])
+        item_size, model_size = np.dot(np.linalg.pinv(matrix), vector)
+
+        # Compute the `batch_size` to use up to `max_memory`
+        optimal_batch_size = (max_memory - model_size) / item_size
+        optimal_batch_size = int(optimal_batch_size)
+
+        return {'batch_size': optimal_batch_size,
+                'item_size': item_size,
+                'model_size': model_size,
+                'table': table}
+
+    def get_memory_utilization(self, batch_size, method='train', inputs=None, targets=None, n=20, frequency=0.05,
+                               time_threshold=3, tail_size=20, std_threshold=0.1):
+        """ For a given `batch_size`, make `inputs` and `targets` and compute memory utilization. """
+        inputs = inputs or self.make_placeholder_data(batch_size, to_device=False) # maybe use to_device=False
```

Comment:
I can't find the method `make_placeholder_data` in this file.

Reply:
If it is should be defined in the child class, please, provide an empty method with such a comment.

Reply:
The `make_placeholder_data` method lies in `TorchModel` class. The main idea of mixins to the `TorchModel` class is to split one huge class into several modules for convenience. So following the example of other mixins to this class, I used child's method in a mixin's method.